Fix for the issue 25366.

thinred opened the pull request:
In particular:
- DH groups of size < 1024 are disabled by default
  (there is only one such group: modp1)
- a new cmdline switch --enable-small-dh-groups and
  SMALL_DH_GROUPS_ENABLE env. variable are introduced;
  they override the default setting and therefore enable
  modp1 group
- the docs & tests are updated

mhdawson commented:
This looks good to me but would like @shigeki  to take a look as well.  If shigeki thinks it looks ok I can create a PR that pulls over nodejs/io.js#1739 and updates to include a check against SMALL_DH_GROUPS_ENABLE.  

shigeki I'd also like your opinion on whether we should add a similar check in crypto.createDiffieHellman(prime_length) as well. ie reject a prime_length smaller than 768 and use the same command line/env variable to allow fallback.

shigeki commented on doc/api/crypto.markdown:
```diff
@@ -425,21 +425,29 @@ expected.
+be explicitly enabled: either via `'--enable-small-dh-groups'` switch to
+node, or by setting the `'ENABLE_SMALL_DH_GROUPS'` environment variable
+to any value.
 
```
I think that this is somehow confusing that only `modp1` leads throwing an error and `modp2` and `modp5` can be used without no message of deprecation whereas the size of 2048 bits is considered deprecated.
Is it better to say that  the group size of less than 1024 bits was deprecated and that of less than 2048 bits is not recommended?

thinred replied:
Ok, let's do it that way.

shigeki commented on src/node_crypto_groups.h:
```diff
@@ -374,13 +382,13 @@ typedef struct {
+  { "modp15", 3072, group_modp15, sizeof(group_modp15) / sizeof(group_modp15[0]), two_generator, 1 },
+  { "modp16", 4096, group_modp16, sizeof(group_modp16) / sizeof(group_modp16[0]), two_generator, 1 },
+  { "modp17", 6144, group_modp17, sizeof(group_modp17) / sizeof(group_modp17[0]), two_generator, 1 },
+  { "modp18", 8192, group_modp18, sizeof(group_modp18) / sizeof(group_modp18[0]), two_generator, 1 },
```
We can use ARRAY_SIZE macro defined in src/node_internals.h here.

thinred replied:
Yes, and 0.12.x uses it here, but I wanted to keep the patch as simple as possible.

shigeki commented on src/node.cc:
```diff
@@ -2566,6 +2566,7 @@ static void PrintHelp() {
          "  --max-stack-size=val set max v8 stack size (bytes)\n"
          "  --enable-ssl2        enable ssl2\n"
          "  --enable-ssl3        enable ssl3\n"
+         "  --enable-small-dh-groups  enable small dh groups (not recommended)\n"
```
Not recommend options are not only `--enable-small-dh-groups` but also `-enable-ssl2` and `-enable-ssl3`. I wonder if we need to show explicitly its recommendation in the help message.

shigeki commented:
Commit title need to be changed to follow https://github.com/joyent/node/blob/master/CONTRIBUTING.md#commit.

shigeki commented:
I put some comments.
Actually I do not have a confidence to make a deprecation of low security key sizes in `crypto.createDiffieHellman` because RSA below 512 bits also needs it. Another security risk exists in md5 and sha1 but we still continue to using them. 
I made the size check in DHE because TLS has a security risk of MiTM but I wonder if we need to extend it for the use not in the wire.

thinred committed:
crypto: discourage and deprecate small DH groups

This fixes issue 25366.

In particular:
    - DH groups of size < 1024 are disabled by default
      (there is only one such group: modp1) and their
      use will throw an exception
    - a new cmdline switch --enable-small-dh-groups and
      the SMALL_DH_GROUPS_ENABLE env. variable are introduced;
      they override the default setting and therefore enable
      modp1 group
    - the documentation & tests are updated

The change has been triggered by the security report "Imperfect Forward
Secrecy: How Diffie-Hellman Fails in Practice" which proved that small,
predefined DH primes groups should not be used.

thinred commented:
I improved the PR.

mhdawson commented:
@shigeki does your last comment apply only to createDiffieHellman or are you wondering if we should be deprecating for getDiffieHellman() ?

shigeki commented:
My comment above is more general for security hardening on crypto module not tls.
After deprecating modp1,2 and 5 in `getDiffieHellman()` here, we still allow  `createDiffieHellman()` with a low key size. In v0.12, we  still allow `createECDH()` with the curve names that have a small size. We can also use 512 bits RSA in `publicEncrypt()`. 
I'm wondering if this fix leads inconsistent security level between crypto functionalities.

thinred commented:
To be clear, only modp1 is deprecated and disabled. modp2 & modp5 work just fine, but are not recommended.
The main security issue here is that calculations necessary to crack the modp1 group (and bigger groups but to a smaller extent) can be put into a precomputation step and then reused very fast. If you generate a *random* DH group with createDiffieHellman and size of 768 bits, you may be *relatively* safe actually, especially if it is a short-term group.
I originally proposed to disable *or* discourage use of small groups. I'm still ok for the latter - maybe an update to the docs is enough?

shigeki commented:
Yes, only modp1 is to be deprecated. I updated and corrected the above comment. 
I agree that modp1 is the most vulnerable for it can solved within a minute during TLS and VPN handshake.  But this crypto module is not only used on wire but also in offline  where they do not have a handshake timeout. Attacker can crack these encrypted data for several days, weeks and months in this case. In this point of this view,  I think there is no reason to have a special treatment only on modp1.

But to be sure, I'm not against deprecation. If we do this, I'm wondering we are going to the way to have `-enable-small-ecdh-curve`, `-enable-small-dh-key`, `-enable-small-rsa-key`  and whatever on Node in future. This does not seem to be smart. Updating docs is one of our choices to put user's responsibility for using low ciphers.

mhdawson commented:
I think we have these options:

1) This patch +  PR that pulls over nodejs/io.js#1739 and updates to include a check against SMALL_DH_GROUPS_ENABLE. 

2) This PR, but exclude deprecation of modp1, + PR that pulls over nodejs/io.js#1739 and updates to include a check against SMALL_DH_GROUPS_ENABLE. (ie include -enable-small-dh-key but only have it apply to use of key in server).

3) PR that pulls over nodejs/io.js#1739 and updates to include a check against SMALL_DH_GROUPS_ENABLE. 

4) Option 1)  + changes to limit usage of small RSA keys and small ecdh curves

5) Option 1, 2, 3 or 4 + rename -enable-small-dh-key to -enable-small-keys which could then be used for future cases like limiting small ecdh or rsa curves or when we want to change the limits set.   We could make it something like -enable-small-keys=0.10.38 which would set size limit based to the limits in the release specified.   This is similar to what is proposed in respect to the removal of RC4. 

6) Do nothing 

I think we need at least

shigeki commented:
@mhdawson Let me confirm that the option is for tls, crypto or both module? This PR is for crypto functionality and nodejs/io.js#1739 is for tls. As I wrote, they have different security  requirements.

mhdawson commented:
@shigeki, I believe we want an option to fall back to previous behaviour for #1739 since it is a breaking change.  I was thinking the option would be for both as I think we want to limit the number of different command line options if at all possible.  If the issue is that we don't believe that we only need to limit the key sizes for tls in order to address logjam then maybe we should just limit for tls and cover the other cases through documentation.

mhdawson commented:
@shigeki would lit make sense to try and find a time that we could talk over hangouts  to come to agreement on what we think we should do ?

shigeki commented:
@mhdawson Sorry, for my late reply. I agree that https://github.com/joyent/node/pull/25514 needs an option and that's no problem. I am going to update my PR to have the option.
For crypto modules, I think we had better to write down some caveats in the doc at first We do not  recommend to use low security dh group and ecdh curve, RSA/DH with small key size and MD5/SHA-1. But there may be use cases for users who have to use them such as to handle old encrypted backup data in offline.

mhdawson commented:
@shigeki. Sounds reasonable to me to do what you suggest in your last comment which I understand to be:

1) update 25514 to support the option to revert ( you are going to update PR to include command line option/env variable)
2) Document recommendations.  Does it make sense for @thinred to update this PR to simply update the documentation with the  the recommendations that they not use `'modp1'`, `'modp2'` and `'modp5' ?

thinred commented:
Any update on this?

shigeki commented:
@thinred Very sorry, I have little time to work on this before July 23 and will be back to do it after that.

shigeki commented:
Thank you for your patience.  https://github.com/shigeki/node/commit/a1e289d70be24d9deda700c1881d3f599882f0c8 is my proposal to update the doc and please take a look at it.  If it is acceptable I will submit a new PR.

thinred commented:
I added two more patches on top of that: https://github.com/thinred/node/tree/c_c
Feel free to squash them, reorder or whatever.

The important change is that it should say "at least" instead of "more than".
Then, nobody will read Caveats at the end of the page, *unless* we link to it. This is what I do in the second commit (well, I also update the example which used modp5).

shigeki commented:
@thinred Thanks for revising my English. Your change and additions are fine with me but I also would like to add one more commit of https://github.com/shigeki/node/commit/fbd32d68726f0aeeb95c8fd76b80dcca9334e6cf to remove `md5` and `sha1` in the example.

@mhdawson We'd better to move to the iojs issue not to miss this after convergence. How do you think about it?

thinred commented:
@shigeki You forgot about this one: https://github.com/thinred/node/commit/8d8d0841a072c379ae41d07fae5f5b476e584bcc :)

mhdawson commented:
We'll want this in 0.10.X and 0.12.X, and everything from 0.12.X should be merged into nodejs/node as part of the convergance work right ?  Is what you are suggesting that once its in 0.10.X and 0.12.X that we should open another issue to get it into io.js master ?

mhdawson commented:
@shigeki I assume you are planning a separate PR for pulling in  nodejs/io.js#1739 with a command line option to revert

shigeki commented:
@mhdawson I'm planning to separating against https://github.com/joyent/node/pull/25514 where a flag should be included.  This PR becomes a doc fix and it should be applied to both iojs and nodejs. After final review in iojs and getting merged, then we can backport it to joyent/node v0.10/v0.12. I fear the description gets inconsistent.

jasnell commented:
@shigeki ... what is the status on this one?

thinred commented:
FYI The current solution is in the branch https://github.com/thinred/node/commits/c_c.

MichalStaruch commented:
@shigeki In my opinion much better solution than ``--enable-small-dh-key`` would be adding something like *.disabledAlgorithms properties used in Java (defined in jre/lib/security/java.security file). It's a set of predefined reasonably safe settings, but it allows to either weaken or strengthen security constraints (depending on your need).

Sample entries from java.security file:
``jdk.certpath.disabledAlgorithms=MD2, MD5, RSA keySize < 1024``
``jdk.tls.disabledAlgorithms=SSLv3, RC4, MD5withRSA, DH keySize < 768``

shigeki commented:
@MichalStaruch Thanks for your comment. Node do not have such a kind of external property files. IMHO, it is not a good idea to have a new parser for it. Anyway, this fix was already made in another PR of https://github.com/nodejs/node/pull/3890. So I'm going to close this.

MichalStaruch commented:
@shigeki It was just an example of a more generic solution, where cryptography constraints are defined by user, not hard-coded - I never said it has to be defined in file :). Similar mechanism (defaults we can override) is already present in node.js, see ``--tls-cipher-list``.
